### PR TITLE
Re-factor apply_filter and add a unit test 

### PR DESCRIPTION
**Pull request by jakubbrindza**

### CODE REVIEW

**Comment by jakubbrindza** [comment]
@izidormatusov please review!
I don't know how to handle the failing Travis-CI tests though.

File: tests/test_filteredtree.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+from liblarch.treenode import _Node
+from liblarch.tree import MainTree
+from liblarch.filters_bank import FiltersBank
+from liblarch.filteredtree import FilteredTree
+
+
+class TestFilteredTree(unittest.TestCase):
+    def setUp(self):
+        self.added_node = 0
+        self.deleted_node = 0
```
Comment (izidormatusov) [inline tests/test_filteredtree.py]: Can you rename this to  be self.added_nodes/self.deleted_nodes? Now I would expect id of that node, not the count

File: tests/test_filteredtree.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        self.assertFalse(self.filtered_tree.is_displayed(node_id="apple"))
+        self.assertTrue(self.filtered_tree.is_displayed(node_id="google"))
+
+        self.assertEqual(1, self.added_node)
+        self.assertEqual(1, self.deleted_node)
+
+    def test_filteredTree_refresh_without_parameters(self):
+        self.filtersbank.add_filter("true_filter", self.true_filter)
+
+        self.added_node, self.deleted_node = 0, 0
```
Comment (izidormatusov) [inline tests/test_filteredtree.py]: Can you put this into a method like self.reset_counters?

File: tests/test_filteredtree.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        self.filtered_tree.set_callback('added', self.added)
+
+    def search_filter(self, node, parameters):
+        return node.get_id() == parameters['node_id']
+
+    def true_filter(self, node):
+        return True
+
+    def test_filteredTree_refresh_with_parameters(self):
+        self.filtersbank.add_filter("search_filter", self.search_filter)
```
Comment (izidormatusov) [inline tests/test_filteredtree.py]: Can you add here asserts that shows that you have both nodes displayed? It's good to test your assumptions

File: tests/test_filteredtree.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        self.filtered_tree.set_callback('deleted', self.deleted)
+        self.filtered_tree.set_callback('added', self.added)
+
+    def search_filter(self, node, parameters):
+        return node.get_id() == parameters['node_id']
+
+    def true_filter(self, node):
+        return True
+
+    def test_filteredTree_refresh_with_parameters(self):
```
Comment (izidormatusov) [inline tests/test_filteredtree.py]: follow BDD style of naming tests

This class tests behaviour of FilteredTree. And your tests should be

- test_refreshes_ever_time_with_parameters
- test_refreshes_only_with_a_new_filter


You should know what you're trying to test based only on your name

File: tests/test_filteredtree.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        self.filtersbank.add_filter("search_filter", self.search_filter)
+
+        self.filtered_tree.apply_filter("search_filter",
+                                        parameters={'node_id': 'apple'})
+        self.assertTrue(self.filtered_tree.is_displayed(node_id="apple"))
+        self.assertFalse(self.filtered_tree.is_displayed(node_id="google"))
+
+        self.assertEqual(1, self.added_node)
+        self.assertEqual(2, self.deleted_node)
```
Comment (izidormatusov) [inline tests/test_filteredtree.py]: Can you add a comment here that when refiltering happens, all nodes are removed and only nodes that stay are added back? To explain why you have 2 delted nodes

File: liblarch/filteredtree.py [hunk 6]
```diff
@@ -620,10 +620,13 @@ def apply_filter(self, filter_name, parameters=None,
                 raise ValueError(
                     "No filter of name {} in the bank".format(filter_name))
 
-        if filter_name not in self.applied_filters:
-            self.applied_filters.append(filter_name)
-            if refresh:
-                self.refilter()
+        filter_predicate = filter_name not in self.applied_filters or\
```
Comment (izidormatusov) [inline liblarch/filteredtree.py]: Simplify this code. Put at the begging of the function:

```python
should_refilter = False
```

When setting parameters, line 618
```
if filt:
   filt.set_parameters(parameters)
   should_refilter = True
```

When adding filter_name into applied_filters, do again `should_refilter=True`

Before return of this function, just do
```python
if should_refilter:
     self.refilter()
```

It would result in much less complicated code

File: tests/test_filteredtree.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        self.assertEqual(0, self.added_node)
+
+    def added(self, node_id, path):
+        self.added_node += 1
+
+    def deleted(self, node_id, path):
+        self.deleted_node += 1
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (izidormatusov) [inline tests/test_filteredtree.py]: this is not needed. Tests should be run by `./run-tests`

File: tests/test_filteredtree.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+class TestFilteredTree(unittest.TestCase):
+    def setUp(self):
+        self.added_node = 0
+        self.deleted_node = 0
+        self.tree = MainTree()
+        self.filtersbank = FiltersBank(self.tree)
+        self.filtered_tree = FilteredTree(self.tree, self.filtersbank)
+        self.tree.add_node(_Node(node_id="apple"), priority="high")
+        self.tree.add_node(_Node(node_id="google"), priority="high")
```
Comment (izidormatusov) [inline tests/test_filteredtree.py]: Can you remove priority argument? It should work even without it. (I'm not 100% sure)

**Comment by izidormatusov** [comment]
If anybody would wonder: Jakub used the main's liblarch repo instead of his own fork for PR so there were 2 Travis CI targets `.../push` (this is if for any branch on this repo) & `.../pr` (this is for PR). Tests were broken on the master (we forgot to list `pep8` as a requirement). Jakub didn't rebase his code based on my fixes on the master and that broke Travis.

**TL;DR** Travis didn't installed `pep8` and Jakub didn't include my fix on master.
